localstack/services/sns/provider.py

```diff
@@ -186,21 +177,39 @@ def publish_batch(
             if entry.get("MessageStructure") == "json":
                 try:
                     message = json.loads(entry.get("Message"))
+                    # Keys in the JSON object that correspond to supported transport protocols must have
+                    # simple JSON string values.
+                    # Non-string values will cause the key to be ignored.
+                    message = {
+                        key: field for key, field in message.items() if isinstance(field, str)
+                    }
                     if "default" not in message:
                         raise InvalidParameterException(
                             "Invalid parameter: Message Structure - No default entry in JSON message body"
                         )
+                    entry["Message"] = message  # noqa
```

Comment:
Out of curiosity, why do we need the `#noqa` here?

Reply:
Because the type of `Message` should be `str`, but I'm already decoding the JSON string so we don't need to downstream, as it needs to be decoded "synchronously" for validation, and once again asynchronously in the publisher when we need to access the data. I could also remove the # noqa and just let the type warning, it might be safer 😅

Reply:
I will leave that up to you, thanks for the clarification :)